refactor: actual immutable pointing

kallewoof opened the pull request:
```C++
const std::shared_ptr<CWallet> wallet = x;
```
means we can not do `wallet = y`, but we can totally do `wallet->DestructiveOperation()`, contrary to what that line looks like.

This PR

* introduces a new convention: always use const shared pointers to `CWallet`s (even when we mutate the pointed-to thing)
* uses `const shared_ptr<const CWallet>` everywhere where wallets are not modified

In the future, this should preferably apply to all shared pointers, not limited to just `CWallet`s.

Both of these serve the same purpose: to dispell the misconception that `const shared_ptr<X>` immutates `X`. It doesn't, and it's dangerous to leave this misconception as is, for obvious reasons.

codedfile commented:
> ```C++
> const std::shared_ptr<CWallet> wallet = x;
> ```
> means we can not do `wallet = y`, but we can totally do `wallet->DestructiveOperation()`, contrary to what that line looks like.
> 
> This PR
> 
> * introduces a new convention: always use const shared pointers to `CWallet`s (even when we mutate the pointed-to thing)
> * uses `shared_ptr<CWallet const>` everywhere where wallets are not modified
> 
> In the future, this should preferably apply to all shared pointers, not limited to just `CWallet`s.
> 
> Both of these serve the same purpose: to dispell the misconception that `const shared_ptr<X>` immutates `X`. It doesn't, and it's dangerous to leave this misconception as is, for obvious reasons.

codedfile commented:
?

jnewbery commented:
I'm a NACK on this. We use the (common) convention of putting `const` to the left of the thing that's const, except in the case of a const pointer to type (`T* const`) where the const needs to go to the right of the `*` to indicate that it's a const pointer rather than a pointer to const. Changing the style just in this case would be inconsistent and confusing.

Hopefully now that this has bitten you once, you'll not get tripped up by it again. FWIW, I think the notation for smart pointers:

- `const std::shared_ptr<T>` "constant shared pointer to mutable T"
- `std::shared_ptr<const T>` "mutable shared pointer to constant T"

is a bit easier to read than for raw pointers:

- `const T*` "mutable pointer to constant T"
- `T* const` "constant pointer to mutable T"

This is also similar for STL containers:

- `const std::vector<T>` - constant vector of Ts
- `std::vector<const T>` - mutable vector of constant Ts

kallewoof commented:
> I'm a NACK on this. We use the (common) convention of putting `const` to the left of the thing that's const, except in the case of a const pointer to type (`T* const`) where the const needs to go to the right of the `*` to indicate that it's a const pointer rather than a pointer to const. Changing the style just in this case would be inconsistent and confusing.

I can easily move the `const` to the left-hand side. We are definitely not consistent with putting it on the left, as you can see all the cases of wallet const shared pointers right now (search for "std::shared_ptr<CWallet> const" in the code base for an example).

That's not what my PR is about though: it's about putting `const` inside the shared pointer, which is a feature that is supported, and that makes the thing inside it immutable, which is what you expect when you see the const sign. You do not expect to be able to call destructive methods or delete things inside a wallet that is marked `const`, which you can do with the above shared pointer.

> Hopefully now that this has bitten you once, you'll not get tripped up by it again.

It is definitely not the first time it's bitten me, personally, but maybe I'm just slow. :)

jnewbery commented on src/wallet/interfaces.cpp:
```diff
@@ -530,7 +530,7 @@ class WalletClientImpl : public WalletClient
     std::unique_ptr<Wallet> createWallet(const std::string& name, const SecureString& passphrase, uint64_t wallet_creation_flags, bilingual_str& error, std::vector<bilingual_str>& warnings) override
     {
-        std::shared_ptr<CWallet> wallet;
+        std::shared_ptr<CWallet> const wallet;
```
Declaring a local `const std::shared_ptr<T> const_ptr;` is pointless. It can't be set to anything after this declaration since it's const.

In fact, this `wallet` variable isn't used, so I think it can just be removed.

kallewoof replied:
Yes, this one just sorta tagged along. I have no idea why it is there, so I didn't touch it beyond letting it tag along with the convention change.

jnewbery reviewed (commented):
> That's not what my PR is about though: it's about putting const inside the shared pointer, which is a feature that is supported, and that makes the thing inside it immutable, which is what you expect when you see the const sign.

Ah, sorry - I misunderstood. I have no objection to making const things const, although I have a strong preference for following the convention of placing `const` on the left side of the type.

kallewoof commented:
> I have a strong preference for following the convention of placing `const` on the left side of the type.

I'm totally fine with putting `const` to the left. However, that's definitely not what is happening in the code, to the point where I ended up putting it on the right here to stay with the code style, despite having no preference about it myself.

kallewoof committed:
refactor: const shared_ptrs

Introduce convention to use const shared pointers everywhere, unless the shared pointer is modified at some point, which it very rarely is.

We want this convention, as it helps alleviate the misconception that a const shared pointer somehow results in a pointer to an immutable object, which is false.

kallewoof committed:
refactor: use CWallet const shared pointers when possible

While const shared_ptr<X> gives us an immutable shared pointer to a mutable X (we can't set it to some other X later), shared_ptr<const X> gives us a shared pointer to an immutable X. Importantly, we can recast shared_ptr<X> into shared_ptr<const X>, but not the other way around. We do this for two reasons: because it makes the code safer to guarantee the wallet is not modified, and because it further dispells the misconception that const shared_ptr<X> gives immutability to X.

kallewoof commented:
Switched to left-hand-`const` in all touched places.

meshcollider commented:
Concept ACK

theStack reviewed (approved):
Code-review ACK 6b98ce32bfcd2b7b93875bbdbdd16b5df47eb94f 🍶 

Based on this PR, I opened a follow-up #22805 which also uses CWallet const shared pointers on the `dumpprivkey` and `dumpwallet` RPC methods.

promag reviewed (commented):
Concept ACK.

6b98ce32bfcd2b7b93875bbdbdd16b5df47eb94f could just change to `std::shared_ptr<const T>` though.

kallewoof commented:
> could just change to `std::shared_ptr<const T>` though.

I'm not sure I follow. Are you suggesting replacing `cons std::shared_ptr<const CWallet>` with `const std::shared_ptr<const T>`? Where does `T` come from?

jaysonmald35 commented:
What are you talking about?

On Sun, Aug 29, 2021, 11:25 PM kallewoof ***@***.***> wrote:

> could just change to std::shared_ptr<const T> though.
>
> I'm not sure I follow. Are you suggesting replacing const
> std::shared_ptr<const CWallet> with const std::shared_ptr<const T>? Where
> does T come from?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/22787#issuecomment-907983973>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AU2W2OK7LGGA5WHKUGBZ2SDT7L2Y3ANCNFSM5CV7JAAA>
> .
>

jaysonmald35 commented:
Don't replace

kallewoof commented:
@promag That makes sense to me, but it isn't clear to the reviewer why some lines are left as `... const ...` and some are changed to `const ... ...`.

practicalswift commented:
Concept ACK

nurinbontot commented:
![IMG_20210906_133552.jpg](https://user-images.githubusercontent.com/83041864/132442301-17e17187-6a91-43d1-af70-7e8eff7dd95d.jpg)

kiminuo commented:
> I'm limiting the scope on this PR to only shared pointers to `CWallet`s to avoid a too big diff.

So should the lines like the following ones be fixed too? I mean to move `const` from the right side to the left side.

```
src\wallet\rpcdump.cpp:122:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcdump.cpp:213:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcdump.cpp:251:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcdump.cpp:339:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcdump.cpp:400:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcdump.cpp:447:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcdump.cpp:528:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcdump.cpp:684:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcdump.cpp:734:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcdump.cpp:1346:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(mainRequest);
src\wallet\rpcdump.cpp:1661:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(main_request);
src\wallet\rpcwallet.cpp:255:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:308:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:354:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:491:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:915:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:988:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:1786:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:1859:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:1912:    std::shared_ptr<CWallet> const wallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:2000:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:2053:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:2097:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:2170:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:2324:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:2586:    std::shared_ptr<CWallet> const wallet = LoadWallet(context, wallet_name, load_on_start, options, status, error, warnings);
src\wallet\rpcwallet.cpp:2673:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:3356:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:3539:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:3678:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:4203:            std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:4330:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:4527:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:4607:    std::shared_ptr<CWallet> const pwallet = GetWalletForJSONRPCRequest(request);
src\wallet\rpcwallet.cpp:4664:            std::shared_ptr<CWallet> const wallet = GetWalletForJSONRPCRequest(request);
```

Produced by:

```powershell
Get-ChildItem -Recurse *.* | Select-String "std::shared_ptr<CWallet>"  | Select-String "const std::shared_ptr<CWallet>" -notMatch
```

kallewoof commented:
> So should the lines like the following ones be fixed too? I mean to move `const` from the right side to the left side.

It's tempting. Right now I'm only changing the lines that I'm touching for other reasons. If others agree, I can ramp it up to include all cases.

theStack reviewed (approved):
re-ACK 52a26e08113005240fd915ce842fe480a2c25fee

Verified via `git range-diff 6b98ce32...52a26e08` that changes since my previous ACK were only rebase-related.

theStack reviewed (approved):
re-ACK 54011e7aa274bdc1b921440cc8b4623aa1e0d89e

Verified via `git range-diff 52a26e08...54011e7a` that changes since my previous ACK were only rebase-related.
